ports/stm/supervisor/port.c

```diff
@@ -34,22 +34,151 @@
 #include "common-hal/busio/UART.h"
-#include "common-hal/pulseio/PWMOut.h"
-#include "common-hal/pulseio/PulseOut.h"
-#include "common-hal/pulseio/PulseIn.h"
 
-#include "stm32f4/clocks.h"
-#include "stm32f4/gpio.h"
+#if defined(STM32F4)
+    #include "common-hal/pulseio/PWMOut.h"
+    #include "common-hal/pulseio/PulseOut.h"
+    #include "common-hal/pulseio/PulseIn.h"
+#endif
 
-#include "stm32f4xx_hal.h"
+#include "clocks.h"
+#include "gpio.h"
+
+//only enable the Reset Handler overwrite for the H7 for now
+#if defined(STM32H7)
+
+// Device memories must be accessed in order.
+#define DEVICE 2
+// Normal memory can have accesses reorder and prefetched.
+#define NORMAL 0
+// Prevents instruction access.
+#define NO_EXECUTION 1
+#define EXECUTION 0
+// Shareable if the memory system manages coherency.
+#define NOT_SHAREABLE 0
+#define SHAREABLE 1
+#define NOT_CACHEABLE 0
+#define CACHEABLE 1
+#define NOT_BUFFERABLE 0
+#define BUFFERABLE 1
+#define NO_SUBREGIONS 0
+
+extern uint32_t _ld_stack_top;
+
+extern uint32_t _ld_d1_ram_bss_start;
+extern uint32_t _ld_d1_ram_bss_size;
+extern uint32_t _ld_d1_ram_data_destination;
+extern uint32_t _ld_d1_ram_data_size;
+extern uint32_t _ld_d1_ram_data_flash_copy;
+extern uint32_t _ld_dtcm_bss_start;
+extern uint32_t _ld_dtcm_bss_size;
+extern uint32_t _ld_dtcm_data_destination;
+extern uint32_t _ld_dtcm_data_size;
+extern uint32_t _ld_dtcm_data_flash_copy;
+extern uint32_t _ld_itcm_destination;
+extern uint32_t _ld_itcm_size;
+extern uint32_t _ld_itcm_flash_copy;
+
+extern void main(void);
+extern void SystemInit(void);
+
+// This replaces the Reset_Handler in startup_*.S and SystemInit in system_*.c.
+__attribute__((used, naked)) void Reset_Handler(void) {
+    __disable_irq();
+    __set_MSP((uint32_t) &_ld_stack_top);
+
+    // TODO: Is any of this commented stuff actually required? 
+
+    /* Disable I cache and D cache */ 
+    // SCB_DisableICache();
+    // SCB_DisableDCache(); // this causes an instant hardfault if used
+
+    // #if ((__FPU_PRESENT == 1) && (__FPU_USED == 1))
+    //   SCB->CPACR |= ((3UL << 10*2) | (3UL << 11*2));    /* set CP10, CP11 Full Access */
+    // #endif /* ((__FPU_PRESENT == 1) && (__FPU_USED == 1)) */
+
+    // /* Disable Systick which might be enabled by bootrom */
+    // if (SysTick->CTRL & SysTick_CTRL_ENABLE_Msk)
+    // {
+    //     SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+    // }
+
+    /* Disable MPU */
+    ARM_MPU_Disable();
+
+    // Copy all of the itcm code to run from ITCM. Do this while the MPU is disabled because we write
+    // protect it.
+    for (uint32_t i = 0; i < ((size_t) &_ld_itcm_size) / 4; i++) {
+        (&_ld_itcm_destination)[i] = (&_ld_itcm_flash_copy)[i];
+    }
+
+    // The first number in RBAR is the region number. When searching for a policy, the region with
+    // the highest number wins. If none match, then the default policy set at enable applies.
+    // TODO: what is the default policy? Where is that set? 
```

Comment:
Default policy is defined by ARM I believe.

Reply:
k